From 9958ba6e4de6562745b4e63c16b1b42f64bfe982 Mon Sep 17 00:00:00 2001
From: Alban Bedel <alban.bedel@avionic-design.de>
Date: Tue, 8 Mar 2016 13:03:55 +0100
Subject: [PATCH] v4l2object: Fix gst_v4l2_object_get_caps_info() for bayer
 format

gst_v4l2_object_get_caps_info() always returned V4L2_PIX_FMT_SBGGR8
for all bayer formats. This is obviously broken if the device use
another ordering. Fix this by properly reading the format parameter.

Signed-off-by: Alban Bedel <alban.bedel@avionic-design.de>
---
 sys/v4l2/gstv4l2object.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 17847e6..a06a1aa 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -1545,7 +1545,17 @@ gst_v4l2_object_get_caps_info (GstV4l2Object * v4l2object, GstCaps * caps,
     } else if (g_str_equal (mimetype, "video/x-vp8")) {
       fourcc = V4L2_PIX_FMT_VP8;
     } else if (g_str_equal (mimetype, "video/x-bayer")) {
-      fourcc = V4L2_PIX_FMT_SBGGR8;
+      const gchar *format = gst_structure_get_string (structure, "format");
+      if (format) {
+	if (!g_ascii_strcasecmp (format, "bggr"))
+	  fourcc = V4L2_PIX_FMT_SBGGR8;
+	else if (!g_ascii_strcasecmp (format, "gbrg"))
+	  fourcc = V4L2_PIX_FMT_SGBRG8;
+	else if (!g_ascii_strcasecmp (format, "grbg"))
+	  fourcc = V4L2_PIX_FMT_SGRBG8;
+	else if (!g_ascii_strcasecmp (format, "rggb"))
+	  fourcc = V4L2_PIX_FMT_SRGGB8;
+      }
     } else if (g_str_equal (mimetype, "video/x-sonix")) {
       fourcc = V4L2_PIX_FMT_SN9C10X;
     } else if (g_str_equal (mimetype, "video/x-pwc1")) {
-- 
2.7.2

